File: java/extensions/semantickernel-planners/src/main/java/com/microsoft/semantickernel/planner/sequentialplanner/SequentialPlanner.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.planner.sequentialplanner;

import com.microsoft.semantickernel.Kernel;
import com.microsoft.semantickernel.SKBuilders;
import com.microsoft.semantickernel.orchestration.SKContext;
import com.microsoft.semantickernel.planner.PlanningException;
import com.microsoft.semantickernel.planner.actionplanner.Plan;
import com.microsoft.semantickernel.semanticfunctions.PromptTemplateConfig;
import com.microsoft.semantickernel.textcompletion.CompletionSKFunction;
import javax.annotation.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Mono;

/** A planner that uses semantic function to create a sequential plan. */
public class SequentialPlanner {

    private static final Logger LOGGER = LoggerFactory.getLogger(SequentialPlanner.class);
    private static final String StopSequence = "<!--";

    // The name to use when creating semantic functions that are restricted from plan creation
    private static final String RestrictedSkillName = "SequentialPlanner_Excluded";

    private final SequentialPlannerRequestSettings config;

    // the function flow semantic function, which takes a goal and creates an xml plan that can be
    // executed
    private final CompletionSKFunction functionFlowFunction;
    private final Kernel kernel;

    public SequentialPlanner(
            Kernel kernel,
            @Nullable SequentialPlannerRequestSettings config,
            @Nullable String prompt) {
        // Verify.NotNull(kernel);

        if (config == null) {
            config = new SequentialPlannerRequestSettings();
        }

        this.config = config.addExcludedSkillName(RestrictedSkillName);

        String promptTemplate;
        if (prompt != null) {
            promptTemplate = prompt;
        } else {
            promptTemplate = EmbeddedResource.read("skprompt.txt");
        }

        this.functionFlowFunction =
                SKBuilders.completionFunctions()
                        .withKernel(kernel)
                        .withPromptTemplate(promptTemplate)
                        .withSkillName(RestrictedSkillName)
                        .withDescription(
                                "Given a request or command or goal generate a step by step plan to"
                                    + " fulfill the request using functions. This ability is also"
                                    + " known as decision making and function flow")
                        .withCompletionConfig(
                                new PromptTemplateConfig.CompletionConfig(
                                        0.0, 0.0, 0.0, 0.0, this.config.getMaxTokens()))
                        .build();

        this.kernel = kernel;
    }

    /**
     * Create a plan for a goal using a default context build on the current kernel.
     *
     * @param goal The goal to create a plan for.
     * @return The plan
     */
    public Mono<Plan> createPlanAsync(String goal) {
        return createPlanAsync(goal, SKBuilders.context().withKernel(kernel).build());
    }

    /**
     * Create a plan for a goal.
     *
     * @param goal The goal to create a plan for.
     * @param context The semantic kernel context
     * @return The plan
     */
    public Mono<Plan> createPlanAsync(String goal, SKContext context) {
        if (goal == null || goal.isEmpty()) {
            throw new PlanningException(
                    PlanningException.ErrorCodes.INVALID_GOAL, "The goal specified is empty");
        }

        try {
            return new DefaultSequentialPlannerSKContext(context)
                    .getFunctionsManualAsync(goal, this.config)
                    .flatMap(
                            relevantFunctionsManual -> {
                                SKContext updatedContext =
                                        context.setVariable(
                                                        "available_functions",
                                                        relevantFunctionsManual)
                                                .update(goal);

                                return functionFlowFunction.invokeAsync(updatedContext, null);
                            })
                    .map(
                            planResult -> {
                                String planResultString = planResult.getResult().trim();

                                LOGGER.debug("Plan result: " + planResultString);

                                Plan plan =
                                        SequentialPlanParser.toPlanFromXml(
                                                planResultString, goal, context.getSkills());
                                return plan;
                            });
        } catch (Exception e) {
            throw new PlanningException(
                    PlanningException.ErrorCodes.INVALID_PLAN,
                    "Plan parsing error, invalid XML",
                    e);
        }
    }
}

File: dotnet/src/Planners/Planners.Core/Sequential/SequentialPlanner.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.SemanticKernel.AI;

namespace Microsoft.SemanticKernel.Planning;

/// <summary>
/// A planner that uses semantic function to create a sequential plan.
/// </summary>
public sealed class SequentialPlanner
{
    private const string StopSequence = "<!-- END -->";
    private const string AvailableFunctionsKey = "available_functions";

    /// <summary>
    /// Initialize a new instance of the <see cref="SequentialPlanner"/> class.
    /// </summary>
    /// <param name="kernel">The <see cref="Kernel"/> containing services, plugins, and other state for use throughout the operation.</param>
    /// <param name="config">The planner configuration.</param>
    public SequentialPlanner(
        Kernel kernel,
        SequentialPlannerConfig? config = null)
    {
        Verify.NotNull(kernel);

        // Set up config with default value and excluded plugins
        this.Config = config ?? new();
        this.Config.ExcludedPlugins.Add(RestrictedPluginName);

        // Set up prompt template
        string promptTemplate = this.Config.GetPromptTemplate?.Invoke() ?? EmbeddedResource.Read("Sequential.skprompt.txt");

        this._functionFlowFunction = kernel.CreateFunctionFromPrompt(
            promptTemplate: promptTemplate,
            description: "Given a request or command or goal generate a step by step plan to " +
                         "fulfill the request using functions. This ability is also known as decision making and function flow",
            executionSettings: new PromptExecutionSettings()
            {
                ExtensionData = new()
                {
                    { "Temperature", 0.0 },
                    { "StopSequences", new[] { StopSequence } },
                    { "MaxTokens", this.Config.MaxTokens },
                }
            });

        this._kernel = kernel;
        this._logger = kernel.LoggerFactory.CreateLogger(this.GetType()) ?? NullLogger.Instance;
    }

    /// <summary>Creates a plan for the specified goal.</summary>
    /// <param name="goal">The goal for which a plan should be created.</param>
    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
    /// <returns>The created plan.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="goal"/> is null.</exception>
    /// <exception cref="ArgumentException"><paramref name="goal"/> is empty or entirely composed of whitespace.</exception>
    /// <exception cref="KernelException">A plan could not be created.</exception>
    public Task<Plan> CreatePlanAsync(string goal, CancellationToken cancellationToken = default)
    {
        Verify.NotNullOrWhiteSpace(goal);

        return PlannerInstrumentation.CreatePlanAsync(
            createPlanAsync: static (SequentialPlanner planner, string goal, CancellationToken cancellationToken) => planner.CreatePlanCoreAsync(goal, cancellationToken),
            planToString: static (Plan plan) => plan.ToSafePlanString(),
            this, goal, this._logger, cancellationToken);
    }

    private async Task<Plan> CreatePlanCoreAsync(string goal, CancellationToken cancellationToken)
    {
        string relevantFunctionsManual = await this._kernel.Plugins.GetFunctionsManualAsync(this.Config, goal, null, cancellationToken).ConfigureAwait(false);

        ContextVariables vars = new(goal)
        {
            [AvailableFunctionsKey] = relevantFunctionsManual
        };

        FunctionResult planResult = await this._kernel.InvokeAsync(this._functionFlowFunction, vars, cancellationToken).ConfigureAwait(false);

        string? planResultString = planResult.GetValue<string>()?.Trim();

        if (string.IsNullOrWhiteSpace(planResultString))
        {
            throw new KernelException(
                "Unable to create plan. No response from Function Flow function. " +
                $"\nGoal:{goal}\nFunctions:\n{relevantFunctionsManual}");
        }

        var getFunctionCallback = this.Config.GetFunctionCallback ?? this._kernel.Plugins.GetFunctionCallback();

        Plan plan;
        try
        {
            plan = planResultString!.ToPlanFromXml(goal, getFunctionCallback, this.Config.AllowMissingFunctions);
        }
        catch (KernelException e)
        {
            throw new KernelException($"Unable to create plan for goal with available functions.\nGoal:{goal}\nFunctions:\n{relevantFunctionsManual}", e);
        }

        if (plan.Steps.Count == 0)
        {
            throw new KernelException($"Not possible to create plan for goal with available functions.\nGoal:{goal}\nFunctions:\n{relevantFunctionsManual}");
        }

        return plan;
    }

    private SequentialPlannerConfig Config { get; }

    private readonly Kernel _kernel;
    private readonly ILogger _logger;

    /// <summary>
    /// the function flow semantic function, which takes a goal and creates an xml plan that can be executed
    /// </summary>
    private readonly KernelFunction _functionFlowFunction;

    /// <summary>
    /// The name to use when creating semantic functions that are restricted from plan creation
    /// </summary>
    private const string RestrictedPluginName = "SequentialPlanner_Excluded";
}
